### PR TITLE
MOSIP-25355

### PR DESCRIPTION
**Pull request by dhanendra06**

### CODE REVIEW

File: registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java [hunk 1]
```diff
@@ -354,6 +353,9 @@ private CredentialRequestDto getCredentialRequestDto(String regId) {
 		credentialRequestDto.setIssuer(env.getProperty("mosip.registration.processor.issuer"));
 
 		credentialRequestDto.setEncryptionKey(generatePin());
+		additionalAttributes.put("credential_id",regId);
```
Comment (vishwa-vyom) [inline registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java]: I believe we are calling this method passing the VID, do we want to set the vid as credential id ?

File: registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java [hunk 2]
```diff
@@ -345,6 +343,7 @@ public MessageDTO process(MessageDTO object) {
 
 	private CredentialRequestDto getCredentialRequestDto(String regId) {
```
Comment (vishwa-vyom) [inline registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java]: We need to support 2 credential request based on configuration, 1 for print and 1 for digital card service

File: registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java [hunk 3]
```diff
@@ -354,6 +353,9 @@ private CredentialRequestDto getCredentialRequestDto(String regId) {
 		credentialRequestDto.setIssuer(env.getProperty("mosip.registration.processor.issuer"));
 
 		credentialRequestDto.setEncryptionKey(generatePin());
+		additionalAttributes.put("credential_id",regId);
+		additionalAttributes.put("template",env.getProperty("mosip.registration.processor.uincard.template"));
```
Comment (vishwa-vyom) [inline registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java]: This property will give back the template name or full template itself ?  I don't see this property in config PR

Comment (vishwa-vyom) [reply registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java]: also handle the null check or get it as @value with default value as something

File: registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java (hunk 1 above)
Comment (vishwa-vyom) [reply registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java]: also the AID#PDF convention is not followed

Comment (vishwa-vyom) [reply registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java]: Also I believe credential id should in credentialRequestDto, not as additional attributes, this has to be /v2 API

**Review by vishwa-vyom — CHANGES_REQUESTED** [review]

File: registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java (hunk 1 above)
Comment (dhanendra06) [reply registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java]: taken care

File: registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java (hunk 2 above)
Comment (dhanendra06) [reply registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java]: taken care

File: registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java (hunk 3 above)
Comment (dhanendra06) [reply registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java]: taken care

File: registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java [hunk 4]
```diff
@@ -6,9 +6,7 @@
 import java.security.SecureRandom;
 import java.time.LocalDateTime;
 import java.time.format.DateTimeFormatter;
-import java.util.ArrayList;
-import java.util.List;
-import java.util.Random;
+import java.util.*;
```
Comment (loganathan-sekaran) [inline registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java]: Organize import, avoid using wildcard in imports

File: registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java [hunk 5]
```diff
@@ -222,16 +222,28 @@ public MessageDTO process(MessageDTO object) {
 			}
 			else {
 			String vid = getVid(uin);
-			CredentialRequestDto credentialRequestDto = getCredentialRequestDto(vid);
 			requestWrapper.setId(env.getProperty("mosip.registration.processor.credential.request.service.id"));
-			requestWrapper.setRequest(credentialRequestDto);
+			String issuers=	env.getProperty("mosip.registration.processor.issuer");
```
Comment (loganathan-sekaran) [inline registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java]: Make the property name as a constant

File: registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java (hunk 4 above)
Comment (loganathan-sekaran) [inline registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java]: Organize import, avoid using wildcard in imports

File: registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java [hunk 6]
```diff
@@ -222,16 +222,28 @@ public MessageDTO process(MessageDTO object) {
...
 			requestWrapper.setRequesttime(localdatetime);
 			requestWrapper.setVersion("1.0");
-			responseWrapper = (ResponseWrapper<?>) restClientService.postApi(ApiName.CREDENTIALREQUEST, null, null,
-					requestWrapper, ResponseWrapper.class, MediaType.APPLICATION_JSON);
+			for (String key : issuers.split(SEMICOLON)) {
+				CredentialRequestDto credentialRequestDto=null;
+				String[] parts = key.split(COLON, 3);
+				credentialRequestDto = getCredentialRequestDto(regId,parts[0],parts[1],parts[2]);
+				requestWrapper.setRequest(credentialRequestDto);
+				if(parts[0].equalsIgnoreCase("mpartner-default-print")){
```
Comment (loganathan-sekaran) [inline registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java]: The partner ID should not be hardcoded. Please take it from configuration.

File: registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java [hunk 7]
```diff
@@ -222,16 +222,28 @@ public MessageDTO process(MessageDTO object) {
...
+						.parse(DateUtils.getUTCCurrentDateTimeString(env.getProperty(DATETIME_PATTERN)), format);
 			requestWrapper.setRequesttime(localdatetime);
 			requestWrapper.setVersion("1.0");
-			responseWrapper = (ResponseWrapper<?>) restClientService.postApi(ApiName.CREDENTIALREQUEST, null, null,
-					requestWrapper, ResponseWrapper.class, MediaType.APPLICATION_JSON);
+			for (String key : issuers.split(SEMICOLON)) {
+				CredentialRequestDto credentialRequestDto=null;
+				String[] parts = key.split(COLON, 3);
+				credentialRequestDto = getCredentialRequestDto(regId,parts[0],parts[1],parts[2]);
+				requestWrapper.setRequest(credentialRequestDto);
```
Comment (loganathan-sekaran) [inline registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java]: Please add comment over the if condition what is the condition is about.

File: registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java [hunk 8]
```diff
@@ -222,16 +222,28 @@ public MessageDTO process(MessageDTO object) {
...
+				if(parts[0].equalsIgnoreCase("mpartner-default-print")){
+					responseWrapper = (ResponseWrapper<?>) restClientService.postApi(ApiName.CREDENTIALREQUEST, null, null,
+							requestWrapper, ResponseWrapper.class, MediaType.APPLICATION_JSON);
+				}else{
+					List<String> pathsegments=new ArrayList<>();
+					pathsegments.add(regId+"#PDF");
```
Comment (loganathan-sekaran) [inline registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java]: What is this #PDF suffix is about? please add a comment on the purpose of that. Also make it as a constant.

File: registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java [hunk 9]
```diff
@@ -343,17 +355,20 @@ public MessageDTO process(MessageDTO object) {
...
+		Map<String, Object> additionalAttributes=new HashMap<>();
 
-		credentialRequestDto.setCredentialType(env.getProperty("mosip.registration.processor.credentialtype"));
+		credentialRequestDto.setCredentialType(credentialType);
 		credentialRequestDto.setEncrypt(encrypt);
 
 		credentialRequestDto.setId(regId);
 
-		credentialRequestDto.setIssuer(env.getProperty("mosip.registration.processor.issuer"));
+		credentialRequestDto.setIssuer(env.getProperty(issuerId));
```
Comment (loganathan-sekaran) [inline registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java]: I think the entire statement `env.getProperty("mosip.registration.processor.issuer")` needs to be replaced with the `issuerId` . We should not call env.getProperty with the issuerId.

File: registration-processor/post-processor/registration-processor-printing-stage/src/main/resources/bootstrap.properties [hunk 10]
```diff
@@ -22,4 +22,9 @@ eventbus.port=5722
...
 mosip.regproc.eventbus.kafka.max.poll.records=${mosip.regproc.printing.eventbus.kafka.max.poll.records}
 mosip.regproc.eventbus.kafka.poll.frequency=${mosip.regproc.printing.eventbus.kafka.poll.frequency}
-mosip.regproc.eventbus.kafka.group.id=${mosip.regproc.printing.eventbus.kafka.group.id}
\ No newline at end of file
+mosip.regproc.eventbus.kafka.group.id=${mosip.regproc.printing.eventbus.kafka.group.id}
+mosip.registration.processor.digitalcard.credentialtype=PDFCard
+mosip.registration.processor.print.credentialtype=euin
+mosip.registration.processor.digitalcard.template=RPR_UIN_CARD_TEMPLATE
+mosip.registration.processor.print.template=RPR_UIN_CARD_TEMPLATE
+mosip.registration.processor.issuer=mpartner-default-print:${mosip.registration.processor.print.credentialtype}:${mosip.registration.processor.print.template};mpartner-default-digitalcard:${mosip.registration.processor.digitalcard.credentialtype}:${mosip.registration.processor.digitalcard.template}
```
Comment (loganathan-sekaran) [inline registration-processor/post-processor/registration-processor-printing-stage/src/main/resources/bootstrap.properties:31]: Add detailed description about the property with an example.

**Review by loganathan-sekaran — CHANGES_REQUESTED** [review]

File: registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java [hunk 11]
```diff
@@ -343,17 +355,20 @@ public MessageDTO process(MessageDTO object) {
...
 
 		credentialRequestDto.setEncryptionKey(generatePin());
+		additionalAttributes.put("template",template);
```
Comment (loganathan-sekaran) [inline registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java]: I think this is not template itself, but template code, so please re-name the additional attribute name and variable accordingly.

**Review by loganathan-sekaran — CHANGES_REQUESTED** [review]

File: registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java [hunk 12]
```diff
@@ -222,16 +222,28 @@ public MessageDTO process(MessageDTO object) {
...
-					.parse(DateUtils.getUTCCurrentDateTimeString(env.getProperty(DATETIME_PATTERN)), format);
+						.parse(DateUtils.getUTCCurrentDateTimeString(env.getProperty(DATETIME_PATTERN)), format);
 			requestWrapper.setRequesttime(localdatetime);
 			requestWrapper.setVersion("1.0");
-			responseWrapper = (ResponseWrapper<?>) restClientService.postApi(ApiName.CREDENTIALREQUEST, null, null,
-					requestWrapper, ResponseWrapper.class, MediaType.APPLICATION_JSON);
+			for (String key : issuers.split(SEMICOLON)) {
+				CredentialRequestDto credentialRequestDto=null;
+				String[] parts = key.split(COLON, 3);
+				credentialRequestDto = getCredentialRequestDto(regId,parts[0],parts[1],parts[2]);
```
Comment (loganathan-sekaran) [inline registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java]: Add a length check (it should not be less than 3) before accessing the elements in array.

**Review by loganathan-sekaran — CHANGES_REQUESTED** [review]

File: registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java (hunk 12 above)
Comment (dhanendra06) [reply registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java]: taken care

File: registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java (hunk 11 above)
Comment (dhanendra06) [reply registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java]: taken care

File: registration-processor/post-processor/registration-processor-printing-stage/src/main/resources/bootstrap.properties (hunk 10 above)
Comment (dhanendra06) [reply registration-processor/post-processor/registration-processor-printing-stage/src/main/resources/bootstrap.properties:31]: taken care

File: registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java (hunk 9 above)
Comment (dhanendra06) [reply registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java]: taken care

File: registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java (hunk 8 above)
Comment (dhanendra06) [reply registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java]: taken care

File: registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java (hunk 7 above)
Comment (dhanendra06) [reply registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java]: taken care

File: registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java (hunk 6 above)
Comment (dhanendra06) [reply registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java]: taken care

File: registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java (hunk 4 above)
Comment (dhanendra06) [reply registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java]: taken care

File: registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java (hunk 5 above)
Comment (dhanendra06) [reply registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java]: taken care

File: registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java (hunk 4 above)
Comment (dhanendra06) [reply registration-processor/post-processor/registration-processor-printing-stage/src/main/java/io/mosip/registration/processor/print/stage/PrintingStage.java]: taken care

**Review by loganathan-sekaran — APPROVED** [review]

**Review by loganathan-sekaran — APPROVED** [review]

**Review by vishwa-vyom — APPROVED** [review]
Approving based Loga's review

**Pull request merged**